File: 01 CLASE ENTERO/PY/Cadena.py
class Cadena:
    def __init__(self, cadena):
        self.cadena = cadena

    def get_cadena(self):  # Obtiene el valor asignado a la cadena
        return self.cadena

    # Manipulación de cadenas
    def concatenar(self, otra_cadena):  # Unir una cadena con otra
        self.cadena += otra_cadena

    def invertir(self):  # Invertir los caracteres de la cadena
        self.cadena = self.cadena[::-1]

    # Comparaciones
    def es_vacia(self):  # Compara si la cadena no tiene caracteres
        return len(self.cadena) == 0

    def contiene(self, subcadena):  # Compara si la subcadena existe dentro de la cadena
        return subcadena in self.cadena

    def comienza_con(self, prefijo):  # Compara si la cadena comienza con el prefijo "caracter"
        return self.cadena.startswith(prefijo)

    def termina_con(self, sufijo):  # Compara si la cadena termina con el sufijo "caracter"
        return self.cadena.endswith(sufijo)

    # Otras operaciones
    def longitud(self):  # Obtiene la cantidad de caracteres que tiene la cadena
        return len(self.cadena)

    def a_mayusculas(self):  # Obtiene la cadena transformada en MAYÚSCULA
        return self.cadena.upper()

    def a_minusculas(self):  # Obtiene la cadena transformada en minúscula
        return self.cadena.lower()


# Ejemplo de uso
cadena = Cadena('')

def ejecutar(metodo, valor_input=''):
    if metodo == 'get_cadena':
        return f'Cadena: {cadena.get_cadena()}'
    elif metodo == 'concatenar':
        cadena.concatenar(valor_input)
        return f'Cadena actual: {cadena.get_cadena()}'
    elif metodo == 'invertir':
        cadena.invertir()
        return f'Cadena invertida: {cadena.get_cadena()}'
    elif metodo == 'es_vacia':
        return f'¿Es vacía?: {cadena.es_vacia()}'
    elif metodo == 'contiene':
        return f'¿Contiene "{valor_input}"?: {cadena.contiene(valor_input)}'
    elif metodo == 'comienza_con':
        return f'¿Comienza con "{valor_input}"?: {cadena.comienza_con(valor_input)}'
    elif metodo == 'termina_con':
        return f'¿Termina con "{valor_input}"?: {cadena.termina_con(valor_input)}'
    elif metodo == 'longitud':
        return f'Longitud: {cadena.longitud()}'
    elif metodo == 'a_mayusculas':
        return f'A mayúsculas: {cadena.a_mayusculas()}'
    elif metodo == 'a_minusculas':
        return f'A minúsculas: {cadena.a_minusculas()}'
    else:
        return 'Método no reconocido.'

# Ejemplo de llamada a la función
resultado = ejecutar('concatenar', 'Hola')
print(resultado)  # Cadena actual: Hola
resultado = ejecutar('invertir')
print(resultado)  # Cadena invertida: aloH
resultado = ejecutar('a_mayusculas')
print(resultado)  # A mayúsculas: HOLA

File: 02 CLASE CADENA/Js/Cadena.js
class Cadena {
  constructor(cadena) {
    this.cadena = cadena;
  }

  getCadena() {
    return this.cadena;
  }

  concatenar(otraCadena) {
    this.cadena += otraCadena;
  }

  invertir() {
    let cadenaInvertida = '';
    for (let i = this.cadena.length - 1; i >= 0; i--) {
      cadenaInvertida += this.cadena[i];
    }
    this.cadena = cadenaInvertida;
  }

  esVacia() {
    return this.cadena.length === 0;
  }

  contiene(subcadena) {
    return this.cadena.includes(subcadena);
  }

  comienzaCon(prefijo) {
    return this.cadena.startsWith(prefijo);
  }

  terminaCon(sufijo) {
    return this.cadena.endsWith(sufijo);
  }

  longitud() {
    return this.cadena.length;
  }

  aMayusculas() {
    return this.cadena.toUpperCase();
  }

  aMinusculas() {
    return this.cadena.toLowerCase();
  }

  aRomano() {
    const romanos = {
        M: 1000,
        CM: 900,
        D: 500,
        CD: 400,
        C: 100,
        XC: 90,
        L: 50,
        XL: 40,
        X: 10,
        IX: 9,
        V: 5,
        IV: 4,
        I: 1
    };
    let numero = parseInt(this.cadena);
    if (isNaN(numero)) {
        return "No es un número válido.";
    }
    let resultado = '';
    for (let clave in romanos) {
        while (numero >= romanos[clave]) {
            resultado += clave;
            numero -= romanos[clave];
        }
    }
    return resultado;
  }

  generarTexto(nPalabras) {
      const caracteres = 'abcdefghijklmnopqrstuvwxyz0123456789';
      let resultado = '';
      
      for (let i = 0; i < nPalabras; i++) {
          let palabra = '';
          let longitudPalabra = Math.floor(Math.random() * 10) + 1;
          for (let j = 0; j < longitudPalabra; j++) {
              palabra += caracteres.charAt(Math.floor(Math.random() * caracteres.length));
          }
          resultado += i < nPalabras - 1 ? palabra + ' ' : palabra;
      }
      
      return resultado;
  }

  aTexto() {
    const numero = parseInt(this.cadena); // Convertir la cadena a número
    if (isNaN(numero) || numero < 0 || numero > 255) {
        return "Por favor, introduce un número entero positivo entre 0 y 255.";
    }
    return String.fromCharCode(numero); // Convertir el número a su representación ASCII
  }
}


// Instancia de Cadena
let cadena = new Cadena('');

// Función ejecutar para conectar con los botones HTML
function ejecutar(metodo) {
  const input = document.getElementById('inputCadena');
  const resultadoDiv = document.getElementById('resultado');
  const valorInput = input.value;
  cadena.cadena = valorInput;

  switch (metodo) {
      case 'getCadena':
          resultadoDiv.innerText = `Cadena: ${cadena.getCadena()}`;
          break;
      case 'concatenar':
          cadena.concatenar(valorInput);
          resultadoDiv.innerText = `Cadena actual: ${cadena.getCadena()}`;
          break;
      case 'invertir':
          cadena.invertir();
          resultadoDiv.innerText = `Cadena invertida: ${cadena.getCadena()}`;
          break;
      case 'esVacia':
          resultadoDiv.innerText = `¿Es vacía?: ${cadena.esVacia()}`;
          break;
      case 'contiene':
          resultadoDiv.innerText = `¿Contiene '${valorInput}'?: ${cadena.contiene(valorInput)}`;
          break;
      case 'comienzaCon':
          resultadoDiv.innerText = `¿Comienza con '${valorInput}'?: ${cadena.comienzaCon(valorInput)}`;
          break;
      case 'terminaCon':
          resultadoDiv.innerText = `¿Termina con '${valorInput}'?: ${cadena.terminaCon(valorInput)}`;
          break;
      case 'longitud':
          resultadoDiv.innerText = `Longitud: ${cadena.longitud()}`;
          break;
      case 'aMayusculas':
          resultadoDiv.innerText = `A mayúsculas: ${cadena.aMayusculas()}`;
          break;
      case 'aMinusculas':
          resultadoDiv.innerText = `A minúsculas: ${cadena.aMinusculas()}`;
          break;
      case 'aRomano':
          resultadoDiv.innerText = `Números Romanos: ${cadena.aRomano()}`;
          break;
      case 'generarTexto':
          const numPalabras = parseInt(valorInput);
          if (isNaN(numPalabras)) {
              resultadoDiv.innerText = `Por favor, introduce un número válido para generar texto.`;
          } else {
              resultadoDiv.innerText = `Texto generado: ${cadena.generarTexto(numPalabras)}`;
          }
          break;
      case 'aTexto':
          const asciiInput = parseInt(valorInput);
          if (isNaN(asciiInput) || asciiInput < 0 || asciiInput > 255) {
              resultadoDiv.innerText = `Por favor, introduce un número entero positivo entre 0 y 255.`;
          } else {
              resultadoDiv.innerText = `Carácter ASCII: ${cadena.aTexto()}`;
          }
          break;
      default:
          resultadoDiv.innerText = 'Método no reconocido.';
          break;
  }
}
